// @minko-fe/react-hook/dist/lib/useSessionStorageSetState/index.cjs
'use strict';

var lodashPro = require('@minko-fe/lodash-pro');
var ahooks = require('ahooks');
var react = require('react');

var __defProp = Object.defineProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
function createUseStorageState(getStorage) {
  function useStorageState(key, options = {}) {
    let storage;
    const {
      onError = (e) => {
        console.error(e);
      }
    } = options;
    try {
      storage = getStorage();
    } catch (err) {
      onError(err);
    }
    const serializer = (value) => {
      if (options.serializer) {
        return options.serializer(value);
      }
      return JSON.stringify(value);
    };
    const deserializer = (value) => {
      if (options.deserializer) {
        return options.deserializer(value);
      }
      return JSON.parse(value);
    };
    function getStoredValue() {
      try {
        const raw = storage == null ? void 0 : storage.getItem(key);
        if (raw) {
          return deserializer(raw);
        }
      } catch (e) {
        onError(e);
      }
      if (lodashPro.isFunction(options.defaultValue)) {
        return options.defaultValue();
      }
      return options.defaultValue;
    }
    const [state, setState] = react.useState(getStoredValue);
    ahooks.useUpdateEffect(() => {
      setState(getStoredValue());
    }, [key]);
    const updateState = (value) => {
      const currentState = lodashPro.isFunction(value) ? value(state) : value;
      setState(currentState);
      if (lodashPro.isUndefined(currentState)) {
        storage == null ? void 0 : storage.removeItem(key);
      } else {
        try {
          storage == null ? void 0 : storage.setItem(key, serializer(currentState));
        } catch (e) {
          console.error(e);
        }
      }
    };
    return [state, ahooks.useMemoizedFn(updateState)];
  }
  return useStorageState;
}

// src/useSetState/index.ts
function reduceWaterfallPlugins(plugins, pluginName) {
  return (state, prevState) => {
    return plugins.reduce((acc, p) => {
      var _a, _b;
      if (p[pluginName]) {
        acc = (_b = (_a = p[pluginName]) == null ? void 0 : _a.call(p, acc, prevState)) != null ? _b : acc;
      }
      return acc;
    }, state);
  };
}
function useSetStateImpl(initialState, options, plugins) {
  const pluginImpls = plugins.map((plugin) => plugin(options));
  const [state, setState] = react.useState(() => {
    const value = lodashPro.isFunction(initialState) ? initialState() : initialState;
    return reduceWaterfallPlugins(pluginImpls, "onInit")(value, value);
  });
  const stateRef = ahooks.useLatest(state);
  const setMergeState = ahooks.useMemoizedFn((patch) => {
    const prevState = stateRef.current;
    const newState = lodashPro.isFunction(patch) ? patch(prevState) : patch;
    const currentState = newState ? __spreadValues(__spreadValues({}, prevState), newState) : prevState;
    const value = reduceWaterfallPlugins(pluginImpls, "onSetState")(currentState, prevState);
    stateRef.current = value;
    setState(value);
  });
  return [state, setMergeState];
}
var useStoragePlugin = (option) => {
  const { storage } = option;
  const useStorageState = createUseStorageState(() => lodashPro.isBrowser() ? storage.api : void 0);
  const [state, setState] = useStorageState(storage.key, storage.options);
  return {
    onInit: () => {
      return state;
    },
    onSetState: (state2) => {
      setState(state2);
    }
  };
};
function useSetState(initialState, options, plugins) {
  const builtInPlugins = [];
  if (options == null ? void 0 : options.storage) {
    builtInPlugins.push(useStoragePlugin);
  }
  return useSetStateImpl(initialState, options || {}, [...[], ...builtInPlugins]);
}

// src/useSessionStorageSetState/index.ts
function useSessionStorageSetState(key, options) {
  return useSetState(options.defaultValue, {
    storage: {
      api: sessionStorage,
      key,
      options
      // Explicitly cast options to Options<T | undefined>
    }
  });
}

exports.useSessionStorageSetState = useSessionStorageSetState;
