// @minko-fe/react-hook/dist/es/useUrlState.js
import { useUpdate, useMemoizedFn } from 'ahooks';
import querystring from 'query-string';
import { useRef, useMemo } from 'react';
import { useLocation, useNavigate } from 'react-router-dom';

var __defProp = Object.defineProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
const baseParseConfig = {
  parseNumbers: false,
  parseBooleans: false
};
const baseStringifyConfig = {
  skipNull: false,
  skipEmptyString: false
};
const useUrlState = (initialState, options) => {
  const { navigateMode = "push", parseOptions, stringifyOptions } = options || {};
  const mergedParseOptions = __spreadValues(__spreadValues({}, baseParseConfig), parseOptions);
  const mergedStringifyOptions = __spreadValues(__spreadValues({}, baseStringifyConfig), stringifyOptions);
  const location = useLocation();
  const navigate = useNavigate();
  const update = useUpdate();
  const initialStateRef = useRef(typeof initialState === "function" ? initialState() : initialState || {});
  const queryFromUrl = useMemo(() => {
    return querystring.parse(location.search, mergedParseOptions);
  }, [location.search]);
  const targetQuery = useMemo(
    () => __spreadValues(__spreadValues({}, initialStateRef.current), queryFromUrl),
    [queryFromUrl]
  );
  const setState = (s) => {
    const newQuery = typeof s === "function" ? s(targetQuery) : s;
    update();
    navigate(
      {
        hash: location.hash,
        search: querystring.stringify(__spreadValues(__spreadValues({}, queryFromUrl), newQuery), mergedStringifyOptions) || "?"
      },
      {
        replace: navigateMode === "replace",
        state: location.state
      }
    );
  };
  return [targetQuery, useMemoizedFn(setState)];
};

export { useUrlState };
