// @minko-fe/react-hook/dist/lib/index.cjs
'use strict';

var ReactUse = require('react-use');
var ahooks = require('ahooks');
var React = require('react');
var lodashPro = require('@minko-fe/lodash-pro');
var useGlobalStyle = require('use-global-style');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var ReactUse__namespace = /*#__PURE__*/_interopNamespace(ReactUse);
var React__default = /*#__PURE__*/_interopDefault(React);

var __defProp = Object.defineProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
var EventEmitter = class {
  constructor() {
    __publicField(this, "subscriptions", /* @__PURE__ */ new Set());
    __publicField(this, "emit", (val) => {
      for (const subscription of this.subscriptions) {
        subscription(val);
      }
    });
    __publicField(this, "useSubscription", (callback) => {
      const callbackRef = React.useRef();
      callbackRef.current = callback;
      React.useEffect(() => {
        function subscription(val) {
          if (callbackRef.current) {
            callbackRef.current(val);
          }
        }
        this.subscriptions.add(subscription);
        return () => {
          this.subscriptions.delete(subscription);
        };
      }, []);
    });
  }
};
function createUseStorageState(getStorage) {
  function useStorageState(key, options = {}) {
    let storage;
    const {
      onError = (e) => {
        console.error(e);
      }
    } = options;
    try {
      storage = getStorage();
    } catch (err) {
      onError(err);
    }
    const serializer = (value) => {
      if (options.serializer) {
        return options.serializer(value);
      }
      return JSON.stringify(value);
    };
    const deserializer = (value) => {
      if (options.deserializer) {
        return options.deserializer(value);
      }
      return JSON.parse(value);
    };
    function getStoredValue() {
      try {
        const raw = storage == null ? void 0 : storage.getItem(key);
        if (raw) {
          return deserializer(raw);
        }
      } catch (e) {
        onError(e);
      }
      if (lodashPro.isFunction(options.defaultValue)) {
        return options.defaultValue();
      }
      return options.defaultValue;
    }
    const [state, setState] = React.useState(getStoredValue);
    ahooks.useUpdateEffect(() => {
      setState(getStoredValue());
    }, [key]);
    const updateState = (value) => {
      const currentState = lodashPro.isFunction(value) ? value(state) : value;
      setState(currentState);
      if (lodashPro.isUndefined(currentState)) {
        storage == null ? void 0 : storage.removeItem(key);
      } else {
        try {
          storage == null ? void 0 : storage.setItem(key, serializer(currentState));
        } catch (e) {
          console.error(e);
        }
      }
    };
    return [state, ahooks.useMemoizedFn(updateState)];
  }
  return useStorageState;
}
function useControlledState(option) {
  const { defaultValue, value, onChange, beforeValue, postValue } = option;
  const isControlled = Object.prototype.hasOwnProperty.call(option, "value") && typeof value !== "undefined";
  const initialValue = React.useMemo(() => {
    if (isControlled) {
      return value;
    }
    if (defaultValue !== void 0) {
      return lodashPro.isFunction(defaultValue) ? defaultValue() : defaultValue;
    }
  }, []);
  const stateRef = React.useRef(initialValue);
  if (isControlled) {
    stateRef.current = value;
  }
  const previousState = ahooks.usePrevious(stateRef.current);
  if (postValue) {
    const post = postValue(stateRef.current, previousState);
    if (post) {
      stateRef.current = post;
    }
  }
  const update = ahooks.useUpdate();
  function triggerChange(newValue) {
    let r = lodashPro.isFunction(newValue) ? newValue(stateRef.current) : newValue;
    if (beforeValue) {
      const before = beforeValue(r, stateRef.current);
      if (before) {
        r = before;
      }
    }
    if (onChange) {
      onChange(r, stateRef.current);
    }
    if (!isControlled) {
      stateRef.current = r;
      update();
    }
  }
  return [stateRef.current, ahooks.useMemoizedFn(triggerChange), previousState];
}
function useEffectOnce(cb, deps) {
  const mountRef = React.useRef(false);
  React.useEffect(() => {
    let returnValue;
    if (mountRef.current === false) {
      returnValue = cb();
      mountRef.current = true;
    }
    return returnValue;
  }, [deps]);
}
function useStrictInput(v, onVChange, strictRegExp = /\D/g) {
  const [value, setValue] = useControlledState({
    defaultValue: v,
    value: v,
    onChange: onVChange
  });
  const composedRef = React.useRef(false);
  const onCompositionStart = ahooks.useMemoizedFn(() => {
    composedRef.current = true;
  });
  const onCompositionEnd = ahooks.useMemoizedFn(() => {
    composedRef.current = false;
    setValue(value == null ? void 0 : value.replace(strictRegExp, ""));
  });
  const onChange = ahooks.useMemoizedFn((value2) => {
    let v2 = value2;
    if (!v2) {
      setValue(v2);
      return v2;
    }
    if (!composedRef.current) {
      v2 = value2 == null ? void 0 : value2.replace(strictRegExp, "");
    }
    setValue(v2);
    return v2 == null ? void 0 : v2.replace(strictRegExp, "");
  });
  return {
    onCompositionStart,
    onCompositionEnd,
    onChange
  };
}
var MIN_DISTANCE = 10;
function getDirection(x, y) {
  if (x > y && x > MIN_DISTANCE) {
    return "horizontal";
  }
  if (y > x && y > MIN_DISTANCE) {
    return "vertical";
  }
  return "";
}
function useTouch() {
  const startX = React.useRef(0);
  const startY = React.useRef(0);
  const deltaX = React.useRef(0);
  const deltaY = React.useRef(0);
  const offsetX = React.useRef(0);
  const offsetY = React.useRef(0);
  const direction = React.useRef("");
  const isVertical = () => direction.current === "vertical";
  const isHorizontal = () => direction.current === "horizontal";
  const reset = () => {
    deltaX.current = 0;
    deltaY.current = 0;
    offsetX.current = 0;
    offsetY.current = 0;
    direction.current = "";
  };
  const start = (event) => {
    reset();
    startX.current = event.touches[0].clientX;
    startY.current = event.touches[0].clientY;
  };
  const move = (event) => {
    const touch = event.touches[0];
    deltaX.current = touch.clientX < 0 ? 0 : touch.clientX - startX.current;
    deltaY.current = touch.clientY - startY.current;
    offsetX.current = Math.abs(deltaX.current);
    offsetY.current = Math.abs(deltaY.current);
    if (!direction.current) {
      direction.current = getDirection(offsetX.current, offsetY.current);
    }
  };
  return {
    move,
    start,
    reset,
    startX,
    startY,
    deltaX,
    deltaY,
    offsetX,
    offsetY,
    direction,
    isVertical,
    isHorizontal
  };
}

// src/useScrollParent/dom/getTargetElement.ts
function getTargetElement(target, defaultElement) {
  if (!target) {
    return defaultElement;
  }
  let targetElement;
  if (typeof target === "function") {
    targetElement = target();
  } else if ("current" in target) {
    targetElement = target.current;
  } else {
    targetElement = target;
  }
  return targetElement;
}

// src/useScrollParent/index.ts
var overflowScrollReg = /scroll|auto/i;
var defaultRoot = lodashPro.isBrowser() ? window : void 0;
function isElement(node) {
  const ELEMENT_NODE_TYPE = 1;
  return node.tagName !== "HTML" && node.tagName !== "BODY" && node.nodeType === ELEMENT_NODE_TYPE;
}
function getScrollParent(el, root = defaultRoot) {
  if (root === void 0) {
    root = window;
  }
  let node = el;
  while (node && node !== root && isElement(node)) {
    const { overflowY } = window.getComputedStyle(node);
    if (overflowScrollReg.test(overflowY)) {
      if (node.tagName !== "BODY") {
        return node;
      }
      const htmlOverflowY = window.getComputedStyle(node.parentNode).overflowY;
      if (overflowScrollReg.test(htmlOverflowY)) {
        return node;
      }
    }
    node = node.parentNode;
  }
  return root;
}
function useScrollParent(el) {
  const [scrollParent, setScrollParent] = React.useState();
  React.useEffect(() => {
    if (el) {
      const element = getTargetElement(el);
      setScrollParent(getScrollParent(element));
    }
  }, []);
  return scrollParent;
}

// src/useLockScroll/index.ts
var supportsPassive = false;
if (lodashPro.isBrowser()) {
  try {
    const opts = {};
    Object.defineProperty(opts, "passive", {
      get() {
        supportsPassive = true;
      }
    });
    window.addEventListener("test-passive", null, opts);
  } catch (e) {
  }
}
var totalLockCount = 0;
var BODY_LOCK_CLASS = "rc-overflow-hidden";
function useLockScroll(rootRef, shouldLock = false) {
  const touch = useTouch();
  const onTouchMove = (event) => {
    touch.move(event);
    const direction = touch.deltaY.current > 0 ? "10" : "01";
    const el = getScrollParent(event.target, rootRef.current);
    if (!el) return;
    const { scrollHeight, offsetHeight, scrollTop } = el;
    let status = "11";
    if (scrollTop === 0) {
      status = offsetHeight >= scrollHeight ? "00" : "01";
    } else if (scrollTop + offsetHeight >= scrollHeight) {
      status = "10";
    }
    if (status !== "11" && touch.isVertical() && !(Number.parseInt(status, 2) & Number.parseInt(direction, 2))) {
      if (event.cancelable) {
        event.preventDefault();
      }
    }
  };
  const lock = () => {
    document.addEventListener("touchstart", touch.start);
    document.addEventListener("touchmove", onTouchMove, supportsPassive ? { passive: false } : false);
    if (!totalLockCount) {
      document.body.classList.add(BODY_LOCK_CLASS);
    }
    totalLockCount++;
  };
  const unlock = () => {
    if (totalLockCount) {
      document.removeEventListener("touchstart", touch.start);
      document.removeEventListener("touchmove", onTouchMove);
      totalLockCount--;
      if (!totalLockCount) {
        document.body.classList.remove(BODY_LOCK_CLASS);
      }
    }
  };
  React.useEffect(() => {
    if (shouldLock) {
      lock();
      return () => {
        unlock();
      };
    }
  }, [shouldLock]);
}
var useLayoutUpdateEffect = (callback, deps) => {
  const firstMountRef = React__default.default.useRef(true);
  ahooks.useIsomorphicLayoutEffect(() => {
    if (!firstMountRef.current) {
      return callback();
    }
  }, deps);
  React.useLayoutEffect(() => {
    firstMountRef.current = false;
    return () => {
      firstMountRef.current = true;
    };
  }, []);
};
function useConstant(init) {
  const ref = React.useRef(null);
  if (ref.current === null) {
    ref.current = init();
  }
  return ref.current;
}
function reduceWaterfallPlugins(plugins, pluginName) {
  return (state, prevState) => {
    return plugins.reduce((acc, p) => {
      var _a, _b;
      if (p[pluginName]) {
        acc = (_b = (_a = p[pluginName]) == null ? void 0 : _a.call(p, acc, prevState)) != null ? _b : acc;
      }
      return acc;
    }, state);
  };
}
function useSetStateImpl(initialState, options, plugins) {
  const pluginImpls = plugins.map((plugin) => plugin(options));
  const [state, setState] = React.useState(() => {
    const value = lodashPro.isFunction(initialState) ? initialState() : initialState;
    return reduceWaterfallPlugins(pluginImpls, "onInit")(value, value);
  });
  const stateRef = ahooks.useLatest(state);
  const setMergeState = ahooks.useMemoizedFn((patch) => {
    const prevState = stateRef.current;
    const newState = lodashPro.isFunction(patch) ? patch(prevState) : patch;
    const currentState = newState ? __spreadValues(__spreadValues({}, prevState), newState) : prevState;
    const value = reduceWaterfallPlugins(pluginImpls, "onSetState")(currentState, prevState);
    stateRef.current = value;
    setState(value);
  });
  return [state, setMergeState];
}
var useStoragePlugin = (option) => {
  const { storage } = option;
  const useStorageState = createUseStorageState(() => lodashPro.isBrowser() ? storage.api : void 0);
  const [state, setState] = useStorageState(storage.key, storage.options);
  return {
    onInit: () => {
      return state;
    },
    onSetState: (state2) => {
      setState(state2);
    }
  };
};
function useSetState(initialState, options, plugins) {
  const builtInPlugins = [];
  if (options == null ? void 0 : options.storage) {
    builtInPlugins.push(useStoragePlugin);
  }
  return useSetStateImpl(initialState, options || {}, [...plugins || [], ...builtInPlugins]);
}

// src/useLocalStorageSetState/index.ts
function useLocalStorageSetState(key, options) {
  return useSetState(options.defaultValue, {
    storage: {
      api: localStorage,
      key,
      options
      // Explicitly cast options to Options<T | undefined>
    }
  });
}

// src/useSessionStorageSetState/index.ts
function useSessionStorageSetState(key, options) {
  return useSetState(options.defaultValue, {
    storage: {
      api: sessionStorage,
      key,
      options
      // Explicitly cast options to Options<T | undefined>
    }
  });
}

// src/index.ts
var {
  useAsync,
  useAsyncFn,
  useAsyncRetry,
  useBeforeUnload,
  useCopyToClipboard,
  useCss,
  useCustomCompareEffect,
  useDefault,
  useError,
  useEvent,
  useFirstMountState,
  useGetSet,
  useGetSetState,
  useHash,
  useHoverDirty,
  useIntersection,
  useLockBodyScroll,
  useMeasure,
  usePageLeave,
  usePermission,
  usePinchZoom,
  usePreviousDistinct,
  useQueue,
  useRaf,
  useRafLoop,
  useScrollbarWidth,
  useScrolling,
  useSearchParam,
  useSlider,
  useSpeech,
  useStartTyping,
  useWindowScroll,
  useWindowSize
} = ReactUse__namespace;

Object.defineProperty(exports, "clearCache", {
  enumerable: true,
  get: function () { return ahooks.clearCache; }
});
Object.defineProperty(exports, "configResponsive", {
  enumerable: true,
  get: function () { return ahooks.configResponsive; }
});
Object.defineProperty(exports, "createUpdateEffect", {
  enumerable: true,
  get: function () { return ahooks.createUpdateEffect; }
});
Object.defineProperty(exports, "useAntdTable", {
  enumerable: true,
  get: function () { return ahooks.useAntdTable; }
});
Object.defineProperty(exports, "useAsyncEffect", {
  enumerable: true,
  get: function () { return ahooks.useAsyncEffect; }
});
Object.defineProperty(exports, "useBoolean", {
  enumerable: true,
  get: function () { return ahooks.useBoolean; }
});
Object.defineProperty(exports, "useClickAway", {
  enumerable: true,
  get: function () { return ahooks.useClickAway; }
});
Object.defineProperty(exports, "useCookieState", {
  enumerable: true,
  get: function () { return ahooks.useCookieState; }
});
Object.defineProperty(exports, "useCountDown", {
  enumerable: true,
  get: function () { return ahooks.useCountDown; }
});
Object.defineProperty(exports, "useCounter", {
  enumerable: true,
  get: function () { return ahooks.useCounter; }
});
Object.defineProperty(exports, "useCreation", {
  enumerable: true,
  get: function () { return ahooks.useCreation; }
});
Object.defineProperty(exports, "useDebounce", {
  enumerable: true,
  get: function () { return ahooks.useDebounce; }
});
Object.defineProperty(exports, "useDebounceEffect", {
  enumerable: true,
  get: function () { return ahooks.useDebounceEffect; }
});
Object.defineProperty(exports, "useDebounceFn", {
  enumerable: true,
  get: function () { return ahooks.useDebounceFn; }
});
Object.defineProperty(exports, "useDeepCompareEffect", {
  enumerable: true,
  get: function () { return ahooks.useDeepCompareEffect; }
});
Object.defineProperty(exports, "useDeepCompareLayoutEffect", {
  enumerable: true,
  get: function () { return ahooks.useDeepCompareLayoutEffect; }
});
Object.defineProperty(exports, "useDocumentVisibility", {
  enumerable: true,
  get: function () { return ahooks.useDocumentVisibility; }
});
Object.defineProperty(exports, "useDrag", {
  enumerable: true,
  get: function () { return ahooks.useDrag; }
});
Object.defineProperty(exports, "useDrop", {
  enumerable: true,
  get: function () { return ahooks.useDrop; }
});
Object.defineProperty(exports, "useDynamicList", {
  enumerable: true,
  get: function () { return ahooks.useDynamicList; }
});
Object.defineProperty(exports, "useEventEmitter", {
  enumerable: true,
  get: function () { return ahooks.useEventEmitter; }
});
Object.defineProperty(exports, "useEventListener", {
  enumerable: true,
  get: function () { return ahooks.useEventListener; }
});
Object.defineProperty(exports, "useEventTarget", {
  enumerable: true,
  get: function () { return ahooks.useEventTarget; }
});
Object.defineProperty(exports, "useExternal", {
  enumerable: true,
  get: function () { return ahooks.useExternal; }
});
Object.defineProperty(exports, "useFavicon", {
  enumerable: true,
  get: function () { return ahooks.useFavicon; }
});
Object.defineProperty(exports, "useFocusWithin", {
  enumerable: true,
  get: function () { return ahooks.useFocusWithin; }
});
Object.defineProperty(exports, "useFullscreen", {
  enumerable: true,
  get: function () { return ahooks.useFullscreen; }
});
Object.defineProperty(exports, "useFusionTable", {
  enumerable: true,
  get: function () { return ahooks.useFusionTable; }
});
Object.defineProperty(exports, "useGetState", {
  enumerable: true,
  get: function () { return ahooks.useGetState; }
});
Object.defineProperty(exports, "useHistoryTravel", {
  enumerable: true,
  get: function () { return ahooks.useHistoryTravel; }
});
Object.defineProperty(exports, "useHover", {
  enumerable: true,
  get: function () { return ahooks.useHover; }
});
Object.defineProperty(exports, "useInViewport", {
  enumerable: true,
  get: function () { return ahooks.useInViewport; }
});
Object.defineProperty(exports, "useInfiniteScroll", {
  enumerable: true,
  get: function () { return ahooks.useInfiniteScroll; }
});
Object.defineProperty(exports, "useInterval", {
  enumerable: true,
  get: function () { return ahooks.useInterval; }
});
Object.defineProperty(exports, "useIsomorphicLayoutEffect", {
  enumerable: true,
  get: function () { return ahooks.useIsomorphicLayoutEffect; }
});
Object.defineProperty(exports, "useKeyPress", {
  enumerable: true,
  get: function () { return ahooks.useKeyPress; }
});
Object.defineProperty(exports, "useLatest", {
  enumerable: true,
  get: function () { return ahooks.useLatest; }
});
Object.defineProperty(exports, "useLocalStorageState", {
  enumerable: true,
  get: function () { return ahooks.useLocalStorageState; }
});
Object.defineProperty(exports, "useLockFn", {
  enumerable: true,
  get: function () { return ahooks.useLockFn; }
});
Object.defineProperty(exports, "useLongPress", {
  enumerable: true,
  get: function () { return ahooks.useLongPress; }
});
Object.defineProperty(exports, "useMap", {
  enumerable: true,
  get: function () { return ahooks.useMap; }
});
Object.defineProperty(exports, "useMemoizedFn", {
  enumerable: true,
  get: function () { return ahooks.useMemoizedFn; }
});
Object.defineProperty(exports, "useMount", {
  enumerable: true,
  get: function () { return ahooks.useMount; }
});
Object.defineProperty(exports, "useMouse", {
  enumerable: true,
  get: function () { return ahooks.useMouse; }
});
Object.defineProperty(exports, "useMutationObserver", {
  enumerable: true,
  get: function () { return ahooks.useMutationObserver; }
});
Object.defineProperty(exports, "useNetwork", {
  enumerable: true,
  get: function () { return ahooks.useNetwork; }
});
Object.defineProperty(exports, "usePagination", {
  enumerable: true,
  get: function () { return ahooks.usePagination; }
});
Object.defineProperty(exports, "usePrevious", {
  enumerable: true,
  get: function () { return ahooks.usePrevious; }
});
Object.defineProperty(exports, "useRafInterval", {
  enumerable: true,
  get: function () { return ahooks.useRafInterval; }
});
Object.defineProperty(exports, "useRafState", {
  enumerable: true,
  get: function () { return ahooks.useRafState; }
});
Object.defineProperty(exports, "useRafTimeout", {
  enumerable: true,
  get: function () { return ahooks.useRafTimeout; }
});
Object.defineProperty(exports, "useReactive", {
  enumerable: true,
  get: function () { return ahooks.useReactive; }
});
Object.defineProperty(exports, "useRequest", {
  enumerable: true,
  get: function () { return ahooks.useRequest; }
});
Object.defineProperty(exports, "useResetState", {
  enumerable: true,
  get: function () { return ahooks.useResetState; }
});
Object.defineProperty(exports, "useResponsive", {
  enumerable: true,
  get: function () { return ahooks.useResponsive; }
});
Object.defineProperty(exports, "useSafeState", {
  enumerable: true,
  get: function () { return ahooks.useSafeState; }
});
Object.defineProperty(exports, "useScroll", {
  enumerable: true,
  get: function () { return ahooks.useScroll; }
});
Object.defineProperty(exports, "useSelections", {
  enumerable: true,
  get: function () { return ahooks.useSelections; }
});
Object.defineProperty(exports, "useSessionStorageState", {
  enumerable: true,
  get: function () { return ahooks.useSessionStorageState; }
});
Object.defineProperty(exports, "useSet", {
  enumerable: true,
  get: function () { return ahooks.useSet; }
});
Object.defineProperty(exports, "useSize", {
  enumerable: true,
  get: function () { return ahooks.useSize; }
});
Object.defineProperty(exports, "useTextSelection", {
  enumerable: true,
  get: function () { return ahooks.useTextSelection; }
});
Object.defineProperty(exports, "useThrottle", {
  enumerable: true,
  get: function () { return ahooks.useThrottle; }
});
Object.defineProperty(exports, "useThrottleEffect", {
  enumerable: true,
  get: function () { return ahooks.useThrottleEffect; }
});
Object.defineProperty(exports, "useThrottleFn", {
  enumerable: true,
  get: function () { return ahooks.useThrottleFn; }
});
Object.defineProperty(exports, "useTimeout", {
  enumerable: true,
  get: function () { return ahooks.useTimeout; }
});
Object.defineProperty(exports, "useTitle", {
  enumerable: true,
  get: function () { return ahooks.useTitle; }
});
Object.defineProperty(exports, "useToggle", {
  enumerable: true,
  get: function () { return ahooks.useToggle; }
});
Object.defineProperty(exports, "useTrackedEffect", {
  enumerable: true,
  get: function () { return ahooks.useTrackedEffect; }
});
Object.defineProperty(exports, "useUnmount", {
  enumerable: true,
  get: function () { return ahooks.useUnmount; }
});
Object.defineProperty(exports, "useUnmountedRef", {
  enumerable: true,
  get: function () { return ahooks.useUnmountedRef; }
});
Object.defineProperty(exports, "useUpdate", {
  enumerable: true,
  get: function () { return ahooks.useUpdate; }
});
Object.defineProperty(exports, "useUpdateEffect", {
  enumerable: true,
  get: function () { return ahooks.useUpdateEffect; }
});
Object.defineProperty(exports, "useUpdateLayoutEffect", {
  enumerable: true,
  get: function () { return ahooks.useUpdateLayoutEffect; }
});
Object.defineProperty(exports, "useVirtualList", {
  enumerable: true,
  get: function () { return ahooks.useVirtualList; }
});
Object.defineProperty(exports, "useWebSocket", {
  enumerable: true,
  get: function () { return ahooks.useWebSocket; }
});
Object.defineProperty(exports, "useWhyDidYouUpdate", {
  enumerable: true,
  get: function () { return ahooks.useWhyDidYouUpdate; }
});
Object.defineProperty(exports, "useGlobalStyle", {
  enumerable: true,
  get: function () { return useGlobalStyle.useGlobalStyle; }
});
exports.EventEmitter = EventEmitter;
exports.createUseStorageState = createUseStorageState;
exports.useAsync = useAsync;
exports.useAsyncFn = useAsyncFn;
exports.useAsyncRetry = useAsyncRetry;
exports.useBeforeUnload = useBeforeUnload;
exports.useConstant = useConstant;
exports.useControlledState = useControlledState;
exports.useCopyToClipboard = useCopyToClipboard;
exports.useCss = useCss;
exports.useCustomCompareEffect = useCustomCompareEffect;
exports.useDefault = useDefault;
exports.useEffectOnce = useEffectOnce;
exports.useError = useError;
exports.useEvent = useEvent;
exports.useFirstMountState = useFirstMountState;
exports.useGetSet = useGetSet;
exports.useGetSetState = useGetSetState;
exports.useHash = useHash;
exports.useHoverDirty = useHoverDirty;
exports.useIntersection = useIntersection;
exports.useLayoutUpdateEffect = useLayoutUpdateEffect;
exports.useLocalStorageSetState = useLocalStorageSetState;
exports.useLockBodyScroll = useLockBodyScroll;
exports.useLockScroll = useLockScroll;
exports.useMeasure = useMeasure;
exports.usePageLeave = usePageLeave;
exports.usePermission = usePermission;
exports.usePinchZoom = usePinchZoom;
exports.usePreviousDistinct = usePreviousDistinct;
exports.useQueue = useQueue;
exports.useRaf = useRaf;
exports.useRafLoop = useRafLoop;
exports.useScrollParent = useScrollParent;
exports.useScrollbarWidth = useScrollbarWidth;
exports.useScrolling = useScrolling;
exports.useSearchParam = useSearchParam;
exports.useSessionStorageSetState = useSessionStorageSetState;
exports.useSetState = useSetState;
exports.useSlider = useSlider;
exports.useSpeech = useSpeech;
exports.useStartTyping = useStartTyping;
exports.useStrictInput = useStrictInput;
exports.useTouch = useTouch;
exports.useWindowScroll = useWindowScroll;
exports.useWindowSize = useWindowSize;
