// @minko-fe/react-hook/dist/lib/useUrlState.cjs
'use strict';

var ahooks = require('ahooks');
var querystring = require('query-string');
var react = require('react');
var reactRouterDom = require('react-router-dom');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var querystring__default = /*#__PURE__*/_interopDefault(querystring);

var __defProp = Object.defineProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var baseParseConfig = {
  parseNumbers: false,
  parseBooleans: false
};
var baseStringifyConfig = {
  skipNull: false,
  skipEmptyString: false
};
var useUrlState = (initialState, options) => {
  const { navigateMode = "push", parseOptions, stringifyOptions } = options || {};
  const mergedParseOptions = __spreadValues(__spreadValues({}, baseParseConfig), parseOptions);
  const mergedStringifyOptions = __spreadValues(__spreadValues({}, baseStringifyConfig), stringifyOptions);
  const location = reactRouterDom.useLocation();
  const navigate = reactRouterDom.useNavigate();
  const update = ahooks.useUpdate();
  const initialStateRef = react.useRef(typeof initialState === "function" ? initialState() : initialState || {});
  const queryFromUrl = react.useMemo(() => {
    return querystring__default.default.parse(location.search, mergedParseOptions);
  }, [location.search]);
  const targetQuery = react.useMemo(
    () => __spreadValues(__spreadValues({}, initialStateRef.current), queryFromUrl),
    [queryFromUrl]
  );
  const setState = (s) => {
    const newQuery = typeof s === "function" ? s(targetQuery) : s;
    update();
    navigate(
      {
        hash: location.hash,
        search: querystring__default.default.stringify(__spreadValues(__spreadValues({}, queryFromUrl), newQuery), mergedStringifyOptions) || "?"
      },
      {
        replace: navigateMode === "replace",
        state: location.state
      }
    );
  };
  return [targetQuery, ahooks.useMemoizedFn(setState)];
};

exports.useUrlState = useUrlState;
