// @minko-fe/react-route/dist/index.cjs
'use strict';

var reactRouterDom = require('react-router-dom');
var lodashPro = require('@minko-fe/lodash-pro');
var React = require('react');
var reactHook = require('@minko-fe/react-hook');
var jsxRuntime = require('react/jsx-runtime');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefault(React);

var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
var Nav = (props) => {
  const { to, replace } = props;
  const nav = reactRouterDom.useNavigate();
  reactHook.useEffectOnce(() => {
    nav(to, { replace });
  }, []);
  return null;
};
var cache = null;
function getDataType(data) {
  return Object.prototype.toString.call(data).match(/\s(\w+)]/)[1];
}
function Guard({
  element,
  meta,
  onRouteBefore,
  onRouteMount,
  onRouteUnMount
}) {
  meta = meta || {};
  reactHook.useEffectOnce(() => {
    onRouteMount == null ? void 0 : onRouteMount(meta);
    return () => {
      onRouteUnMount == null ? void 0 : onRouteUnMount(meta);
    };
  }, []);
  const { pathname } = reactRouterDom.useLocation();
  const navigate = reactRouterDom.useNavigate();
  if (onRouteBefore) {
    if (cache === element) {
      return element;
    }
    const pathRes = onRouteBefore({ pathname, meta });
    if (getDataType(pathRes) === "Promise") {
      pathRes.then((res) => {
        if (res && res !== pathname) {
          navigate(res, { replace: true });
        }
      });
    } else if (pathRes && pathRes !== pathname) {
      element = /* @__PURE__ */ jsxRuntime.jsx(Nav, { to: pathRes, replace: true });
    }
  }
  cache = element;
  return element;
}
var RouterUtil = class {
  constructor(option) {
    __publicField(this, "routes");
    __publicField(this, "onRouteBefore");
    __publicField(this, "onRouteMount");
    __publicField(this, "onRouteUnMount");
    __publicField(this, "suspense");
    this.routes = option.routes || [];
    this.onRouteBefore = option.onRouteBefore;
    this.onRouteMount = option.onRouteMount;
    this.onRouteUnMount = option.onRouteUnMount;
    this.suspense = option.suspense || /* @__PURE__ */ jsxRuntime.jsx("div", {});
  }
  createClientRoutes(routes) {
    const useRoutesList = [];
    const routeList = lodashPro.cloneDeep(routes);
    routeList.forEach((route) => {
      const item = lodashPro.cloneDeep(route);
      if (item.path === void 0) {
        return;
      }
      if (item.redirect) {
        item.element = /* @__PURE__ */ jsxRuntime.jsx(Nav, { to: item.redirect, replace: true });
      } else if (item.component) {
        item.element = this.lazyLoad(item.component, item.meta || {});
      }
      if (item.children) {
        item.children = this.createClientRoutes(item.children);
      }
      useRoutesList.push(this.deleteSelfProperty(item));
    });
    return useRoutesList;
  }
  deleteSelfProperty(r) {
    delete r.redirect;
    delete r.component;
    delete r.meta;
    return r;
  }
  lazyLoad(importFn, meta) {
    const Component = React__default.default.lazy(importFn);
    const lazyElement = /* @__PURE__ */ jsxRuntime.jsx(React__default.default.Suspense, { fallback: this.suspense, children: /* @__PURE__ */ jsxRuntime.jsx(Component, { _meta: meta }) });
    return /* @__PURE__ */ jsxRuntime.jsx(
      Guard,
      {
        element: lazyElement,
        meta,
        onRouteBefore: this.onRouteBefore,
        onRouteMount: this.onRouteMount,
        onRouteUnMount: this.onRouteUnMount
      }
    );
  }
};

// src/createRoutes.tsx
function CreateRoutes({ routes, onRouteBefore, onRouteMount, suspense, render }) {
  const util = new RouterUtil({
    routes,
    onRouteBefore,
    onRouteMount,
    suspense
  });
  const reactRoutes = util.createClientRoutes(routes);
  const elements = reactRouterDom.useRoutes(reactRoutes);
  if (render) {
    return render(elements);
  }
  return elements;
}

exports.CreateRoutes = CreateRoutes;
exports.RouterUtil = RouterUtil;
